### ISSUE TITLE
mWeb - Login - Inconsistency between the mention displayed in chat and LHN preview

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#52584]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.62-2 
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/5223745
**Issue reported by:** Applause Internal Team

## Action Performed:
1.Log in the app
2. Open any conversation
3. In the composer, type '@[concierge@expensify.com'](mailto:concierge@expensify.com')

## Expected Result:
The display name for the mention result is Concierge

## Actual Result:
There is an inconsistency between the mention in the chat and the LHN preview. In the chat the email is displayed, while in LHN, the name is displayed.

## Workaround:
Unknown

## Platforms:

- [ ] Android: Standalone
- [ ] Android: HybridApp
- [x] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/15c9e857-99ea-4fb3-8745-fddba734a250

  </details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by klajdipaja** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The displayed text when the Expensfiy Concierge is mentioned is different between the comment and previewed message in LHN
### What is the root cause of that problem?
The incosistency comes from having two different logics in how the text is constructed for mentioned accounts.
In the chat we use `MentionUserRenderer` component to render the content of the action item when we have a comment that mentions another account. 
In this line we check wether we should display a formated phone number or the display name depending on if the user has logged in with a phone number or email.

https://github.com/Expensify/App/blob/63c8dd2bb44deb2d62e223530fa0d734f1cfba5c/src/components/HTMLEngineProvider/HTMLRenderers/MentionUserRenderer.tsx#L62
But the second part of the condition never validates and is a de facto dead code because of this part of the format phone number function:
https://github.com/Expensify/App/blob/63c8dd2bb44deb2d62e223530fa0d734f1cfba5c/src/libs/LocalePhoneNumber.ts#L26-L28
We are returning there the number that was passed as input when the input was not a valid number.
And since we pass the user login in `LocalePhoneNumber.formatPhoneNumber(user?.login ?? '')` which for email accounts is the actual email we end up having as a display in `mentionDisplayText` the user email for all cases.
In that case, we have another bug where you cannot see the actual display name of the user but their email address.

On the other end the LHN preview uses in `OptionRowLHNData`to build the display data for each row uses this function:
https://github.com/Expensify/App/blob/63c8dd2bb44deb2d62e223530fa0d734f1cfba5c/src/libs/SidebarUtils.ts#L272
Which uses this function to parse the display name from the last message of the report:

https://github.com/Expensify/App/blob/63c8dd2bb44deb2d62e223530fa0d734f1cfba5c/src/libs/Parser.ts#L39-L44

Which returns the display name from the login info of the mentioned account in onyx: https://github.com/Expensify/App/blob/63c8dd2bb44deb2d62e223530fa0d734f1cfba5c/src/libs/Parser.ts#L18

So the inconsistency is from the message using the plain `user.login` info which contains the email address and the LHN showing the display name of the user.

### What changes do you think we should make in order to solve the problem?

Assuming that we want to show the display name in both cases.
We have a few options how to fix the MentionUserRenderer so that it shows the display name:
Option a. 
In the `formatPhoneNumber` return an empty string if the number is not valid:
```
   // do not parse the string, if it doesn't contain the SMS domain and it's not a phone number
    if (number.indexOf(CONST.SMS.DOMAIN) === -1 && !CONST.REGEX.DIGITS_AND_PLUS.test(number)) {
        return '';
    }
```
Easy fix but I'm afraid it might have a regression in one of the many uses that it has 

Option b.
Flip the parts of the condition in the `mentionDisplayText` so that we check the displayName first:
```
        mentionDisplayText = PersonalDetailsUtils.getDisplayNameOrDefault(user) ||  LocalePhoneNumber.formatPhoneNumber(user?.login ?? '');
```
I tested this one and it works well for phone number accounts as well where it shows the full name that the user has created during the on boarding phase.

Option C. Check what the login 
```
If(Str.isSMSLogin(passedPersonalDetails?.login)){
     mentionDisplayText = LocalePhoneNumber.formatPhoneNumber(user?.login ?? '');
}else{
   mentionDisplayText = PersonalDetailsUtils.getDisplayNameOrDefault(user) ||  LocalePhoneNumber.formatPhoneNumber(user?.login ?? '');
}
```
With those we and up having the Display Name from the logged user in the mention but for Concierce we change that explictely here:

https://github.com/Expensify/App/blob/63c8dd2bb44deb2d62e223530fa0d734f1cfba5c/src/libs/PersonalDetailsUtils.ts#L71-L73
Instead of using `Expensify Concierge` we use `Concierge`.

We need to add the same logic for what we show in LHN Preview by changing the mapper in `Parser.ts` to something like this:

```
accountIDToNameMap[personalDetails.accountID] = getDisplayName(personalDetails);
```
where we use this new function:
```
const getDisplayName=(personalDetails) =>{
    if (personalDetails?.accountID === CONST.ACCOUNT_ID.CONCIERGE) {
        return CONST.CONCIERGE_DISPLAY_NAME;
    }
    return personalDetails.login ?? personalDetails.displayName ?? '';
}
```
### What alternative solutions did you explore? (Optional)

**Comment by klajdipaja** [comment]
Updated proposal https://github.com/Expensify/App/issues/52584#issuecomment-2482787909 as the last section mentioend caused a regression. I suggest that change to be done in the BE. Or we align our front end to the same value the BE brings us for the mention name.

**Comment by klajdipaja** [comment]
Updated proposal https://github.com/Expensify/App/issues/52584#issuecomment-2482787909  with another option for the 2nd part of the issue that so far is not causing any regressions.

**Comment by hungvu193** [comment]
My first thought was that this issue should be fixed from Backend. But I'll take a closer look on the proposal above this week if It can change my mind 😄

**Comment by greg-schroeder** [comment]
Thanks @hungvu193 - let me know if does need to be updated to Internal

**Comment by hungvu193** [comment]
We don't format the `lastMessageText` if it includes mention yet, so the `lastMessageText` with mention is currently an message from our BE.

@klajdipaja  If you can provide the code for your `Options Part 2.` without any code warning and clean, I think we can consider it.

**Comment by klajdipaja** [comment]
Updated `Options Part 2` in the proposal https://github.com/Expensify/App/issues/52584#issuecomment-2482787909 with clean code as suggested by @hungvu193.
I have created this draft PR with those changes: https://github.com/Expensify/App/pull/52913
The outcome of that is this:


https://github.com/user-attachments/assets/c95a250d-bef0-4426-82a4-85ea8db3735e

**Comment by hungvu193** [comment]
Thanks I'll take other look today!

**Comment by hungvu193** [comment]
Still under review

**Comment by hungvu193** [comment]
Thanks @klajdipaja for the detailed solution.
After I took a few tests and this issue is only reproducible with report or self DM, not with group chat. I think the best way to handle this is fix it from our BE.


🎀 👀 🎀

**Comment by klajdipaja** [comment]
@hungvu193 what about this part of the proposal.  Is it desired that in the message in chat we display the email instead of `Concierge` ?

> In this line we check wether we should display a formated phone number or the display name depending on if the user has logged in with a phone number or email.
```
 mentionDisplayText = LocalePhoneNumber.formatPhoneNumber(user?.login ?? '') || PersonalDetailsUtils.getDisplayNameOrDefault(user); 
```

> But the second part of the condition never validates and is a de facto dead code because of this part of the format phone number function:
```
 if (number.indexOf(CONST.SMS.DOMAIN) === -1 && !CONST.REGEX.DIGITS_AND_PLUS.test(number)) { 
     return number; 
 } 
```

> We are returning there the number that was passed as input when the input was not a valid number.

**Comment by hungvu193** [comment]
> @hungvu193 what about this part of the proposal. Is it desired that in the message in chat we display the email instead of `Concierge` ?

Yes, I think that's expected.

**Comment by klajdipaja** [comment]
@hungvu193 in that case it seems like a happy accident not like it was intended by the code

**Comment by greg-schroeder** [comment]
bump @youssef-lr on contributor assignment

**Comment by youssef-lr** [comment]
@greg-schroeder does mentioning Concierge do anything in the product right now? I'm not sure this bug is something real world users will run into otherwise and I'm not sure it's worth adding a specific check to avoid it.

**Comment by greg-schroeder** [comment]
I don't think it does. If you're not passionate about it we can close for now until it's encountered by real world users. I'll close, but please reopen if you strongly disagree.
